:seedling: Sync with upstream open-cluster-management-io/cluster-proxy 2026-07-07 (#643)
stolostron/cluster-proxy: CLAUDE.md

- **`.tekton/`** -- Tekton pipeline configs for Konflux CI (MCE component builds)
- **`renovate.json`** -- Renovate bot config extending stolostron/acm-config
- **`sonar-project.properties`** -- SonarQube code analysis config
- **`cmd/Dockerfile.rhtap`** -- RHTAP/Konflux-oriented container build (used by Tekton pipelines)
- **`cmd/pure.Dockerfile`** -- Prow CI container build (used by OpenShift CI prow jobs)
- **`OWNERS`** -- The downstream OWNERS lists the Red Hat ACM team (dhaiducek, tesshuflower,

>   rokej, mikeshng), which is entirely different from the upstream community maintainers. Always

Reviewer: (nit) It seems it'd be more maintainable if the usernames weren't explicitly stated here.

```suggestion
- **`OWNERS`** -- The downstream OWNERS lists the Red Hat ACM team, which is entirely different from the upstream community maintainers. Always
```

  preserve the stolostron version during syncs.

## Dockerfile Landscape

There are four Dockerfiles in this repo. Understanding which is used by which CI system is
important for syncs:

| Dockerfile | Used by | Builder image | Notes |
|---|---|---|---|

> | `cmd/Dockerfile` | Upstream CI; local `make images` (GitHub Actions **disabled** on stolostron) | `golang:<version>` | **Upstream-owned.** Take upstream's version during syncs -- no conflict resolution needed. Upstream downloads vanilla ANP tarball; stolostron does NOT use this file in active CI. |

Reviewer: (nit) I didn't know what ANP is, so it's possible an agent may not know either right offhand. I'd prefer it was spelled out every time as "apiserver-network-proxy" rather than "ANP".

### Step 2: Create a sync branch

```bash

> git checkout stolostron/main

Reviewer: Having a remote called `stolostron` isn't universally true since it's customizable. There might be another way, but I usually parse the `remote -v` output for this:

```suggestion
remote=$(git remote -v | awk '/stolostron.*(push)/ {print $1}')
git checkout "${remote}/main"
```

git merge upstream/main --no-ff
```

Using `--no-ff` ensures a merge commit is created even if a fast-forward were possible,
making the sync point visible in history. Individual upstream commits are preserved with
their original SHAs.

### Step 4: Resolve conflicts

Three files always conflict (see "Expected Merge Conflicts" section above):

```bash
# OWNERS: always keep stolostron version
git checkout --ours OWNERS

# cmd/Dockerfile: always take upstream version
git checkout --theirs cmd/Dockerfile

# third_party/apiserver-network-proxy: keep submodule, restore from HEAD
git checkout HEAD -- third_party/apiserver-network-proxy
```

Then update the submodule to the correct version and branch:

```bash
# Update .gitmodules if the ANP branch has changed (e.g. v0.34.0-patch -> v0.36.0-patch)
# Edit .gitmodules branch field, then:
cd third_party/apiserver-network-proxy
git fetch origin <new-branch>
git checkout <new-tip-commit-sha>

> cd ../..

Reviewer: Git submodules are weird, but you ought to be able to use submodule commands to sync this rather than a fetch/checkout. Had to Google this one 😅 Looks like that `--remote` flag is key for this operation.

```suggestion
git submodule update --init --recursive --remote
```

```

> Also check `go.mod` and `go.sum` for conflicts from dependency version differences:

Reviewer: `go.sum` doesn't need to be checked. Just check `go.mod`, then run `go mod tidy`.

Resolution strategy:

> - For dependencies that exist in both repos: **take upstream's version** (source of truth)

Reviewer: Would a better rule of thumb be to keep the higher version?

changes often need to be mirrored manually.

If the upstream sync includes a **Go version bump** (check `go.mod` for a changed `go` directive),
the downstream-only Dockerfiles must be updated manually -- they are not touched by the merge.

```bash
# Check current Go version in go.mod
go list -m -json go

# Check current versions in downstream Dockerfiles
grep -i "^FROM .* AS builder$" cmd/Dockerfile.rhtap cmd/pure.Dockerfile
```

Update `cmd/Dockerfile.rhtap` (builder image tag pattern is `rhel_<rhel-version>_<goversion>`, e.g.):

```
FROM brew.registry.redhat.io/rh-osbs/openshift-golang-builder:rhel_9_1.26 AS builder
```

Update `cmd/pure.Dockerfile` (builder image tag pattern is `go<goversion>-linux`):

```
FROM registry.ci.openshift.org/stolostron/builder:go1.26-linux AS builder
```

Also update the ANP submodule version comment in both files if the submodule was updated.

Note: `cmd/Dockerfile` and `test/e2e/Dockerfile` are upstream-owned and updated automatically
by the merge. The prow CI config (`openshift/release` repo) may also need its `build_root`
tag updated if the builder image changes -- check

> `ci-operator/config/stolostron/cluster-proxy/stolostron-cluster-proxy-main.yaml`.

Reviewer: This is good, but for the future we can use a `.ci-operator.yaml` file here in the repo so that we don't need to update `openshift/release`. 🙂

Commit these Dockerfile changes separately (signed):

```bash
git add cmd/Dockerfile.rhtap cmd/pure.Dockerfile
git commit -s -S -m "chore: upgrade downstream Dockerfiles to Go <version>"
```

### Step 10: Push and open PR

```bash

> git push origin sync-upstream-$(date +%Y%m%d)

Reviewer: Same with above--we should parse the remote rather than assume it's `origin`. This could be a pattern match like "not stolostron, not OCM-io" or something? Either that or just have the agent push to a new branch in `stolostron`.

git fetch upstream main

# See commits in upstream that aren't in downstream

> git log --oneline upstream/main --not stolostron/main

Reviewer: (Same comment about remotes "upstream" and "stolostron")

# See downstream-only commits

> git log --oneline stolostron/main --not upstream/main

Reviewer: (Same comment about remotes "upstream" and "stolostron")

```

If your `origin` is a personal fork and `stolostron/cluster-proxy` is another remote, substitute
that remote name for `origin` in the `git log` examples so you compare against the downstream

> integration branch you care about.

Reviewer: Ah, here's the note for what I've been talking about. 😬 Still, the `remote -v` parsing could be nice.
